compiler/p/codegen/PPCBinaryEncoding.cpp

```diff
@@ -56,46 +56,374 @@ namespace TR { class Register; }
+   {
+   TR_ASSERT_FATAL_WITH_INSTRUCTION(instr, (val & 0xffu) == val, "0x%x is out-of-range for FLM field", val);
+   *cursor |= val << 17;
+   }
+
+static void fillFieldFXM(TR::Instruction *instr, uint32_t *cursor, uint32_t val)
+   {
+   TR_ASSERT_FATAL_WITH_INSTRUCTION(instr, (val & 0xffu) == val, "0x%x is out-of-range for FXM field", val);
+   *cursor |= val << 12;
+   }
+
+static void fillFieldFXM1(TR::Instruction *instr, uint32_t *cursor, uint32_t val)
+   {
+   TR_ASSERT_FATAL_WITH_INSTRUCTION(instr, populationCount(val), "0x%x is invalid for FXM field, expecting exactly 1 bit set", val);
+   fillFieldFXM(instr, cursor, val);
+   }
+
+static void fillFieldDCM(TR::Instruction *instr, uint32_t *cursor, uint32_t val)
+   {
+   TR_ASSERT_FATAL_WITH_INSTRUCTION(instr, (val & 0x3fu) == val, "0x%x is out-of-range for DCM/DQM field", val);
+   *cursor |= val << 10;
+   }
+
+static void fillFieldSH5(TR::Instruction *instr, uint32_t *cursor, uint32_t val)
+   {
+   TR_ASSERT_FATAL_WITH_INSTRUCTION(instr, (val & 0x1fu) == val, "0x%x is out-of-range for SH(5) field", val);
+   *cursor |= val << 11;
+   }
+
+static void fillFieldSH6(TR::Instruction *instr, uint32_t *cursor, uint32_t val)
+   {
+   TR_ASSERT_FATAL_WITH_INSTRUCTION(instr, (val & 0x3fu) == val, "0x%x is out-of-range for SH(6) field", val);
+   *cursor |= (val & 0x1fu) << 11;
+   *cursor |= (val & 0x20u) >> 4;
+   }
+
+static void fillFieldSIM(TR::Instruction *instr, uint32_t *cursor, uint32_t val)
+   {
+   TR_ASSERT_FATAL_WITH_INSTRUCTION(instr, isValidInSignExtendedField(val, 0x1fu), "0x%x is out-of-range for SIM field", val);
+   *cursor |= (val & 0x1f) << 16;
+   }
+
+static void fillFieldSI(TR::Instruction *instr, uint32_t *cursor, uint32_t val)
+   {
+   TR_ASSERT_FATAL_WITH_INSTRUCTION(instr, isValidInSignExtendedField(val, 0xffffu), "0x%x is out-of-range for SI field", val);
+   *cursor |= val & 0xffff;
+   }
+
+static void fillFieldSI5(TR::Instruction *instr, uint32_t *cursor, uint32_t val)
+   {
+   TR_ASSERT_FATAL_WITH_INSTRUCTION(instr, isValidInSignExtendedField(val, 0x1fu), "0x%x is out-of-range for SI(5) field", val);
+   *cursor |= (val & 0x1f) << 11;
+   }
+
+static void fillFieldUIM(TR::Instruction *instr, uint32_t *cursor, int32_t numBits, uint32_t val)
+   {
+   uint32_t fieldMask = (1u << numBits) - 1;
+   TR_ASSERT_FATAL_WITH_INSTRUCTION(instr, (val & fieldMask) == val, "0x%x is out-of-range for UIM(%d) field", val, numBits);
+   *cursor |= val << 16;
+   }
+
+static void fillFieldUI(TR::Instruction *instr, uint32_t *cursor, uint32_t val)
+   {
+   // TODO: This is a hack until PIC sites are reworked. Currently, the PIC site handling code
+   // assumes that the entire address is in the immediate of the instruction, so we can't chop it to
+   // 16 bits like we should.
+   if (!instr->cg()->comp()->isPICSite(instr))
+      TR_ASSERT_FATAL_WITH_INSTRUCTION(instr, (val & 0xffffu) == val, "0x%x is out-of-range for UI field", val);
```

Comment:
I have impression that, on POWER, we only register the last instruction for PIC site regardless of the whole address. When it is compensated, 1 is or-ed into that instruction (effectively the last chunk of 16-bit).

Reply:
You're right that only the last instruction is actually registered and that's exactly why this assert is a problem on PIC sites. Registering a PIC site happens in `addMetaDataForCodeAddress` during binary encoding [1], which needs to know the full class pointer in order to be able to register it correctly. In order to know the class pointer that needs to be used, we need the full address available at binary encoding time, which prevents us from doing the correct thing by using only the bottom 16 bits of that address as the immediate of the instruction that's set as the PIC site.

The code for registering PIC sites is obviously not designed particularly well, nor is it particularly efficient. It even goes so far as to perform a linear scan of _all PIC sites_ for every instruction that is of a class that could possibly be a PIC site. I'd like to fix that at some point, but this isn't something that can be easily fixed as part of this work, especially since all codegens share some of the infrastructure for keeping track of PIC sites.

[1] https://github.com/eclipse/omr/blob/1679756b7b1b5674cf22a856d78018079e2d2f07/compiler/p/codegen/PPCBinaryEncoding.cpp#L607-L620

Reply:
My point was:  registering didn't need to reconstruct the full address from the registered instruction.  The full address must be available from the node or somewhere. i.e.  the registered instruction's 16bit chunk doesn't matter.  Maybe I missed your point.

Reply:
You're right that the full address should be available _somewhere_, but the existing code relies on that address being in the immediate and changing how `addMetaDataForCodeAddress` determines the address to use requires changes beyond the scope of this PR.

Reply:
if previously only the last instruction in PICsite sequence was registered, how could the full address be determined from that registered instruction's 16-bit immediate?  that is why I said the 16-bit chunk doesn't matter.

Reply:
Ah, I see where the confusion is here. In the past, we never actually validated that the immediate on the `TR::Instruction` actually fit in 16 bits. We would just accept any pointer-sized integer and proceed to only encode the lower 16 bits, ignoring the remainder for encoding purposes. Those upper bits _were_ being used during PIC site registration, so providing the full address as the immediate then having the binary encoder only encode the lower 16 bits was important for functional correctness.

The assert you're looking at here is newly added as part of this refactor and is validating that the immediate fits within the expected range. In most cases, the fix to out-of-range immediates being used was to adjust code that previously assumed that immediates would be implicitly shortened to remove that assumption and only provide the immediate that should _really_ be encoded, since the upper bits were unused. However, for PIC sites, these upper bits were important for the PIC site to be registered correctly, so we need to find a new way to communicate this information to the PIC site registration code before this assert can be enabled on PIC sites.

Reply:
interesting ... i.e. previously intentionally passing the full address as the immediate, instead of typical low16Bits(fullAddress).

Reply:
I am wondering 64bit J9Class pointer though.  Is the instruction immediate 64bit nowadays?  I can imagine 64bit immediate in the future.

Reply:
The sizes of instruction immediates are not currently consistent across the various different instruction classes. In this particular instance, the immediate on a `PPCTrg1Src1ImmInstruction` is pointer-sized. However, `getSourceImmediate()` only returns the lower 32 bits and `getSourceImmPtr()` needs to be used to get the full immediate, meaning the binary encoder currently only ever looks at it as a 32-bit value. Other instruction classes do things differently and none of them are consistent at the moment.

I have a change in progress to restore consistency here and always use `int64_t` for all immediates, though it has hit a couple snags along the way. The current plan is to have that change happen after the phase 3 and phase 4 binary encoder refactors.